packages/kubernetes-1.22/credential-provider-config-yaml

```diff
@@ -11,8 +11,15 @@ providers:
 {{/each}}
     defaultCacheDuration: "{{default "12h" this.cache-duration}}"
     apiVersion: credentialprovider.kubelet.k8s.io/v1alpha1
-{{#if (eq @key "ecr-credential-provider")}}
+{{#if (or (eq @key "ecr-credential-provider") this.env-variables)}}
     env:
+{{#if this.env-variables}}
+{{#each this.env-variables}}
+      - name: {{@key}}
+        value: {{this}}
+{{/each}}
+{{/if}}
+{{#if (eq @key "ecr-credential-provider")}}
       - name: HOME
```

Comment:
If `HOME` and `AWS_PROFILE` appear twice in this list - which can happen because they're set for `ecr-credential-provider` in the template, and also set in settings - does `kubelet` accept the YAML as well-formed or does that cause issues?

Reply:
I was thinking about that possibility too, though now I can't remember if I actually got around to testing out behavior. I will double check and report back on that to be safe. I suspect the environment variable just ends up with whatever is parsed last, but will make sure that really is the case.

Reply:
```sh
$ apiclient set --json '{"aws": {"profile": "ecr"}}'
```

```sh
$ apiclient apply <<EOF
>   [settings.kubernetes.credential-providers.ecr-credential-provider]
>   enabled = true
>   # (optional - defaults to "12h")
>   cache-duration = "30m"
>   image-patterns = [
>     # One or more URL paths to match an image prefix. Supports globbing of subdomains.
>     "*.dkr.ecr.us-east-2.amazonaws.com",
>     "*.dkr.ecr.us-west-2.amazonaws.com"
>   ]
> 
>   [settings.kubernetes.credential-providers.ecr-credential-provider.environment]
>   # The following are not used with ecr-credential-provider, but are provided for illustration
>   "KEY" = "abc123xyz"
>   "GOMAXPROCS" = "2"
>   AWS_PROFILE = "whodis"
>   HOME = "/usr/libexec"
> EOF
```

```json
$ apiclient get settings.kubernetes.credential-providers
{
  "settings": {
    "kubernetes": {
      "credential-providers": {
        "ecr-credential-provider": {
          "cache-duration": "30m",
          "enabled": true,
          "environment": {
            "AWS_PROFILE": "whodis",
            "GOMAXPROCS": "2",
            "HOME": "/usr/libexec",
            "KEY": "abc123xyz"
          },
          "image-patterns": [
            "*.dkr.ecr.us-east-2.amazonaws.com",
            "*.dkr.ecr.us-west-2.amazonaws.com"
          ]
        }
      }
    }
  }
}
```

```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: CredentialProviderConfig
providers:
  - name: ecr-credential-provider
    matchImages:
      - "*.dkr.ecr.us-east-2.amazonaws.com"
      - "*.dkr.ecr.us-west-2.amazonaws.com"
    defaultCacheDuration: "30m"
    apiVersion: credentialprovider.kubelet.k8s.io/v1alpha1
    env:
      - name: AWS_PROFILE
        value: 'whodis'
      - name: GOMAXPROCS
        value: '2'
      - name: HOME
        value: '/usr/libexec'
      - name: KEY
        value: 'abc123xyz'
      - name: HOME
        value: /root
      - name: AWS_PROFILE
        value: ecr
```

```text
# systemctl status kubelet
● kubelet.service - Kubelet
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kubelet.service; enabled; vendor preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kubelet.service.d
             └─dockershim-symlink.conf
             /etc/systemd/system/kubelet.service.d
             └─exec-start.conf
             /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kubelet.service.d
             └─load-ipvs-modules.conf, make-kubelet-dirs.conf, prestart-pull-pause-ctr.conf
     Active: active (running) since Wed 2023-04-05 14:42:28 UTC; 2min 4s ago
```

Kubelet service is happy, template is rendered as expected, and no errors.

Duplicate environment variables mean one or the other will be the last one set, but that's a user configuration issue. The service itself does not have an issue with the extra configuration.